nft-marketplace/marketplace/src/auction.rs

```diff
@@ -0,0 +1,218 @@
+                owner_id: msg::source(),
+                nft_contract_id: *nft_contract_id,
+                ft_contract_id,
+                token_id,
+                price: None,
+                auction: Some(auction),
+                offers: Vec::new(),
+            });
+
+        msg::reply(
+            MarketEvent::AuctionCreated {
+                nft_contract_id: *nft_contract_id,
+                token_id,
+                price: min_price,
+            },
+            0,
+        );
+    }
+
+    /// Settles the auction
+    /// Requirements:
+    /// * The auction must be over
+    /// Arguments:
+    /// * `nft_contract_id`: the NFT contract address
+    /// * `token_id`: the NFT id
+    pub async fn settle_auction(&mut self, nft_contract_id: &ActorId, token_id: U256) {
+        let contract_and_token_id =
+            format!("{}{}", H256::from_slice(nft_contract_id.as_ref()), token_id);
+        let item = self
+            .items
+            .get_mut(&contract_and_token_id)
+            .expect("Item does not exist");
+
+        let auction = item.auction.clone().expect("Auction doesn not exist");
+
+        if auction.ended_at > exec::block_timestamp() {
+            panic!("Auction is not over");
+        }
+
+        let mut bids = auction.bids.unwrap_or_default();
+
+        if !bids.is_empty() {
+            let highest_bid = &bids[bids.len() - 1].clone();
+            bids.pop();
+            for bid in bids {
+                transfer_payment(&exec::program_id(), &bid.id, item.ft_contract_id, bid.price)
+                    .await;
+            }
+            // fee for treasury
+            let treasury_fee = highest_bid.price * self.treasury_fee / 10_000u128;
+            transfer_payment(
+                &exec::program_id(),
+                &self.treasury_id,
+                item.ft_contract_id,
+                treasury_fee,
+            )
+            .await;
+            let payouts = nft_payouts(
+                nft_contract_id,
+                &item.owner_id,
+                highest_bid.price - treasury_fee,
+            )
+            .await;
+            for (account, amount) in payouts.iter() {
+                transfer_payment(&exec::program_id(), account, item.ft_contract_id, *amount).await;
+            }
+
+            item.owner_id = highest_bid.id;
+            // transfer NFT
+            nft_transfer(nft_contract_id, &highest_bid.id, token_id).await;
+            msg::reply(
+                MarketEvent::AuctionSettled {
+                    nft_contract_id: *nft_contract_id,
+                    token_id,
+                    price: highest_bid.price,
+                },
+                0,
+            );
+        } else {
+            msg::reply(
+                MarketEvent::AuctionCancelled {
+                    nft_contract_id: *nft_contract_id,
+                    token_id,
+                },
+                0,
+            );
+        }
+        item.auction = None;
```

Comment:
If I'm not mistaken this variable is not used in either if-else blocks so could be made right after clone:
```rust
        let auction = item.auction.clone().expect("Auction doesn not exist");
        if auction.ended_at > exec::block_timestamp() {
            panic!("Auction is not over");
        }
        
        item.auction = None;
```